File: core/src/main/java/vn/loitp/core/loitp/gallery/photos/GalleryCorePhotosActivity.java
package vn.loitp.core.loitp.gallery.photos;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.animation.OvershootInterpolator;
import android.widget.TextView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import loitp.core.R;
import vn.loitp.core.base.BaseFontActivity;
import vn.loitp.core.common.Constants;
import vn.loitp.core.loitp.gallery.slide.GalleryCoreSlideActivity;
import vn.loitp.core.utilities.LActivityUtil;
import vn.loitp.core.utilities.LDialogUtil;
import vn.loitp.core.utilities.LLog;
import vn.loitp.core.utilities.LSocialUtil;
import vn.loitp.core.utilities.LUIUtil;
import vn.loitp.restapi.flickr.FlickrConst;
import vn.loitp.restapi.flickr.model.photosetgetphotos.Photo;
import vn.loitp.restapi.flickr.service.FlickrService;
import vn.loitp.restapi.restclient.RestClient;
import vn.loitp.views.progressloadingview.avloadingindicatorview.lib.avi.AVLoadingIndicatorView;
import vn.loitp.views.recyclerview.animator.adapters.ScaleInAnimationAdapter;
import vn.loitp.views.recyclerview.animator.animators.SlideInRightAnimator;

public class GalleryCorePhotosActivity extends BaseFontActivity {
    private TextView tvTitle;
    private AVLoadingIndicatorView avLoadingIndicatorView;

    private int currentPage = 0;
    private int totalPage = 1;
    private final int PER_PAGE_SIZE = 100;

    private boolean isLoading;
    private PhotosAdapter photosAdapter;
    private FloatingActionButton btPage;
    private String photosetID;

    private int bkgRootView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        isShowAdWhenExit = false;
        setTransparentStatusNavigationBar();
        PhotosDataCore.getInstance().clearData();
        btPage = findViewById(R.id.bt_page);
        tvTitle = findViewById(R.id.tv_title);
        LUIUtil.setTextShadow(tvTitle);

        avLoadingIndicatorView = findViewById(R.id.av);
        //ImageView ivBkg = (ImageView) findViewById(R.id.iv_bkg);
        //LImageUtil.load(activity, Constants.URL_IMG_2, ivBkg);

        photosetID = getIntent().getStringExtra(Constants.getSK_PHOTOSET_ID());
        final String photosSize = getIntent().getStringExtra(Constants.getSK_PHOTOSET_SIZE());
        LLog.d(TAG, "photosetID " + photosetID);
        LLog.d(TAG, "photosSize " + photosSize);

        bkgRootView = getIntent().getIntExtra(Constants.getBKG_ROOT_VIEW(), Constants.getNOT_FOUND());
        LLog.d(TAG, "bkgRootView " + bkgRootView);
        if (bkgRootView == Constants.getNOT_FOUND()) {
            getRootView().setBackgroundColor(ContextCompat.getColor(activity, R.color.colorPrimary));
        } else {
            getRootView().setBackgroundResource(bkgRootView);
        }

        int totalPhotos;
        try {
            totalPhotos = Integer.parseInt(photosSize);
        } catch (final Exception e) {
            showDialogError(getString(R.string.err_unknow));
            return;
        }

        if (totalPhotos == 0) {
            showDialogError(getString(R.string.err_unknow));
            return;
        }

        if (totalPhotos % PER_PAGE_SIZE == 0) {
            totalPage = totalPhotos / PER_PAGE_SIZE;
        } else {
            totalPage = totalPhotos / PER_PAGE_SIZE + 1;
        }

        currentPage = totalPage;
        //LLog.d(TAG, "total page " + totalPage);
        //LLog.d(TAG, "currentPage " + currentPage);

        final RecyclerView recyclerView = findViewById(R.id.recyclerView);

        final SlideInRightAnimator animator = new SlideInRightAnimator(new OvershootInterpolator(1f));
        animator.setAddDuration(1000);
        recyclerView.setItemAnimator(animator);

        final int column = 2;

        recyclerView.setLayoutManager(new GridLayoutManager(activity, column));
        recyclerView.setHasFixedSize(true);
        photosAdapter = new PhotosAdapter(activity, column, new PhotosAdapter.Callback() {
            @Override
            public void onClick(final Photo photo, final int pos) {
                //LLog.d(TAG, "onClick " + photo.getWidthO() + "x" + photo.getHeightO());
                final Intent intent = new Intent(activity, GalleryCoreSlideActivity.class);
                intent.putExtra(Constants.INSTANCE.getSK_PHOTO_ID(), photo.getId());
                intent.putExtra(Constants.getBKG_ROOT_VIEW(), bkgRootView);
                startActivity(intent);
                LActivityUtil.tranIn(activity);
            }

            @Override
            public void onLongClick(final Photo photo, final int pos) {
                LSocialUtil.share(activity, photo.getUrlO());
            }
        });
        //recyclerView.setAdapter(albumAdapter);
        final ScaleInAnimationAdapter scaleAdapter = new ScaleInAnimationAdapter(photosAdapter);
        scaleAdapter.setDuration(1000);
        scaleAdapter.setInterpolator(new OvershootInterpolator());
        scaleAdapter.setFirstOnly(true);
        recyclerView.setAdapter(scaleAdapter);

        LUIUtil.setPullLikeIOSVertical(recyclerView);

        photosetsGetPhotos(photosetID);

        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                if (!recyclerView.canScrollVertically(1)) {
                    if (!isLoading) {
                        currentPage--;
                        LLog.d(TAG, "last item ->>> ");
                        photosetsGetPhotos(photosetID);
                    }
                }
            }
        });

        btPage.setOnClickListener(v -> {
            //LLog.d(TAG, "onClick " + currentPage + "/" + totalPage);
            showListPage();
        });
    }

    private void showListPage() {
        int size = totalPage;
        final String arr[] = new String[size];
        for (int i = 0; i < size; i++) {
            arr[i] = "Page " + (totalPage - i);
        }
        LDialogUtil.showDialogList(activity, "Select page", arr, position -> {
            currentPage = totalPage - position;
            LLog.d(TAG, "showDialogList onClick position " + position + ", -> currentPage: " + currentPage);
            PhotosDataCore.getInstance().clearData();
            updateAllViews();
            photosetsGetPhotos(photosetID);
        });
    }

    @Override
    protected boolean setFullScreen() {
        return false;
    }

    @Override
    protected String setTag() {
        return "TAG" + getClass().getSimpleName();
    }

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_gallery_core_photos;
    }

    private void photosetsGetPhotos(@NonNull final String photosetID) {
        if (isLoading) {
            LLog.d(TAG, "isLoading true -> return");
            return;
        }
        LLog.d(TAG, "is calling photosetsGetPhotos " + currentPage + "/" + totalPage);
        isLoading = true;
        avLoadingIndicatorView.smoothToShow();
        final FlickrService service = RestClient.createService(FlickrService.class);
        final String method = FlickrConst.METHOD_PHOTOSETS_GETPHOTOS;
        final String apiKey = FlickrConst.API_KEY;
        final String userID = FlickrConst.USER_KEY;
        if (currentPage <= 0) {
            LLog.d(TAG, "currentPage <= 0 -> return");
            currentPage = 0;
            avLoadingIndicatorView.smoothToHide();
            isLoading = false;
            return;
        }
        final String primaryPhotoExtras = FlickrConst.PRIMARY_PHOTO_EXTRAS_1;
        final String format = FlickrConst.FORMAT;
        final int nojsoncallback = FlickrConst.NO_JSON_CALLBACK;

        compositeDisposable.add(service.photosetsGetPhotos(method, apiKey, photosetID, userID, primaryPhotoExtras, PER_PAGE_SIZE, currentPage, format, nojsoncallback)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(wrapperPhotosetGetPhotos -> {
                    //LLog.d(TAG, "photosetsGetPhotos " + currentPage + "/" + totalPage);
                    final String s = wrapperPhotosetGetPhotos.getPhotoset().getTitle() + " (" + currentPage + "/" + totalPage + ")";
                    tvTitle.setText(s);
                    final List<Photo> photoList = wrapperPhotosetGetPhotos.getPhotoset().getPhoto();
                    PhotosDataCore.getInstance().addPhoto(photoList);
                    updateAllViews();
                    avLoadingIndicatorView.smoothToHide();
                    btPage.setVisibility(View.VISIBLE);
                    isLoading = false;
                }, e -> {
                    LLog.e(TAG, "onFail " + e.toString());
                    handleException(e);
                    avLoadingIndicatorView.smoothToHide();
                    isLoading = true;
                }));
    }

    private void updateAllViews() {
        if (photosAdapter != null) {
            photosAdapter.notifyDataSetChanged();
        }
    }
}

File: core/src/main/java/com/core/helper/gallery/photos/GalleryCorePhotosActivity.kt
package com.core.helper.gallery.photos

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.R
import com.annotation.IsSwipeActivity
import com.annotation.LogTag
import com.core.base.BaseFontActivity
import com.core.common.Constants
import com.core.helper.gallery.slide.GalleryCoreSlideActivity
import com.core.utilities.LActivityUtil
import com.core.utilities.LDialogUtil
import com.core.utilities.LSocialUtil
import com.core.utilities.LValidateUtil
import com.restapi.flickr.FlickrConst
import com.restapi.flickr.model.photosetgetphotos.Photo
import com.restapi.flickr.service.FlickrService
import com.restapi.restclient.RestClient
import com.views.layout.swipeback.SwipeBackLayout
import com.views.setSafeOnClickListener
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import jp.wasabeef.recyclerview.adapters.SlideInBottomAnimationAdapter
import kotlinx.android.synthetic.main.l_activity_flickr_gallery_core_photos.*

@LogTag("GalleryCorePhotosActivity")
@IsSwipeActivity(true)
class GalleryCorePhotosActivity : BaseFontActivity() {
    private var currentPage = 0
    private var totalPage = 1
    private var isLoading = false
    private var photosAdapter: PhotosAdapter? = null
    private var photosetID: String? = null

    override fun setLayoutResourceId(): Int {
        return R.layout.l_activity_flickr_gallery_core_photos
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        PhotosDataCore.instance.clearData()

        photosetID = intent.getStringExtra(Constants.SK_PHOTOSET_ID)
        val photosSize = intent.getStringExtra(Constants.SK_PHOTOSET_SIZE)

        val totalPhotos = try {
            photosSize?.toInt() ?: 0
        } catch (e: Exception) {
            e.printStackTrace()
            showDialogError(getString(R.string.err_unknow))
            return
        }
        if (totalPhotos == 0) {
            showDialogError(getString(R.string.err_unknow))
            return
        }
        totalPage = if (totalPhotos % Constants.PER_PAGE_SIZE == 0) {
            totalPhotos / Constants.PER_PAGE_SIZE
        } else {
            totalPhotos / Constants.PER_PAGE_SIZE + 1
        }
        currentPage = totalPage

        val column = 2
        recyclerView.layoutManager = GridLayoutManager(this, column)
//        recyclerView.setHasFixedSize(true)
        photosAdapter = PhotosAdapter(
                callback = object : PhotosAdapter.Callback {
                    override fun onClick(photo: Photo, pos: Int) {
                        val intent = Intent(this@GalleryCorePhotosActivity, GalleryCoreSlideActivity::class.java)
                        intent.putExtra(Constants.SK_PHOTO_ID, photo.id)
                        startActivity(intent)
                        LActivityUtil.tranIn(this@GalleryCorePhotosActivity)
                    }

                    override fun onLongClick(photo: Photo, pos: Int) {
                        LSocialUtil.share(activity = this@GalleryCorePhotosActivity, msg = photo.urlO)
                    }
                })

        photosAdapter?.let {
//            val animAdapter = AlphaInAnimationAdapter(it)
//            val animAdapter = ScaleInAnimationAdapter(it)
            val animAdapter = SlideInBottomAnimationAdapter(it)
//            val animAdapter = SlideInLeftAnimationAdapter(it)
//            val animAdapter = SlideInRightAnimationAdapter(it)

            animAdapter.setDuration(1000)
//            animAdapter.setInterpolator(OvershootInterpolator())
            animAdapter.setFirstOnly(true)
            recyclerView.adapter = animAdapter
        }
        //LUIUtil.setPullLikeIOSVertical(recyclerView)

        photosetsGetPhotos(photosetID = photosetID)
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                if (!recyclerView.canScrollVertically(1)) {
                    if (!isLoading) {
                        currentPage--
                        photosetsGetPhotos(photosetID = photosetID)
                    }
                }
            }
        })
        btPage.setSafeOnClickListener {
            showListPage()
        }

        swipeBackLayout.setSwipeBackListener(object : SwipeBackLayout.OnSwipeBackListener {
            override fun onViewPositionChanged(mView: View?, swipeBackFraction: Float, swipeBackFactor: Float) {
            }

            override fun onViewSwipeFinished(mView: View?, isEnd: Boolean) {
                if (isEnd) {
                    finish()
                    LActivityUtil.transActivityNoAnimation(this@GalleryCorePhotosActivity)
                }
            }
        })

        LValidateUtil.isValidPackageName()
    }

    private fun showListPage() {
        val size = totalPage
        val arr = arrayOfNulls<String>(size)
        for (i in 0 until size) {
            arr[i] = "Page " + (totalPage - i)
        }
        LDialogUtil.showDialogList(
                context = this,
                title = "Select page",
                arr = arr,
                onClick = { position ->
                    currentPage = totalPage - position
                    PhotosDataCore.instance.clearData()
                    updateAllViews()
                    photosetsGetPhotos(photosetID)
                }
        )
    }

    private fun photosetsGetPhotos(photosetID: String?) {
        if (isLoading || photosetID.isNullOrEmpty()) {
            return
        }
        isLoading = true
        indicatorView.smoothToShow()
        val service = RestClient.createService(FlickrService::class.java)
        val method = FlickrConst.METHOD_PHOTOSETS_GETPHOTOS
        val apiKey = FlickrConst.API_KEY
        val userID = FlickrConst.USER_KEY
        if (currentPage <= 0) {
            currentPage = 0
            indicatorView.smoothToHide()
            isLoading = false
            return
        }
        val primaryPhotoExtras = FlickrConst.PRIMARY_PHOTO_EXTRAS_1
        val format = FlickrConst.FORMAT
        val noJsonCallBack = FlickrConst.NO_JSON_CALLBACK
        compositeDisposable.add(service.getPhotosetPhotos(
                method = method,
                apiKey = apiKey,
                photosetId = photosetID,
                userId = userID,
                extras = primaryPhotoExtras,
                perPage = Constants.PER_PAGE_SIZE,
                page = currentPage,
                format = format,
                noJsonCallback = noJsonCallBack)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ wrapperPhotosetGetPhotos ->
                    val s = wrapperPhotosetGetPhotos?.photoset?.title + " (" + currentPage + "/" + totalPage + ")"
                    tvTitle.text = s
                    wrapperPhotosetGetPhotos?.photoset?.photo?.let {
                        it.shuffle()
                        PhotosDataCore.instance.addPhoto(it)
                    }
                    updateAllViews()
                    indicatorView.smoothToHide()
                    btPage.visibility = View.VISIBLE
                    isLoading = false
                }) { e: Throwable ->
                    handleException(e)
                    indicatorView.smoothToHide()
                    isLoading = true
                })
    }

    private fun updateAllViews() {
        photosAdapter?.notifyDataSetChanged()
    }
}
